### PR TITLE
common: fix a racing in PerfCounters::perf_counter_data_any_d::read_avg

### PR DESCRIPTION
**Pull request by luhp**
PerfCounters::perf_counter_data_any_d::read_avg is not safe, maybe get wrong data.

### CODE REVIEW

**Comment by tchaikov** [comment]
retest this please.

**Comment by luhp** [comment]
Here is my code to confirm the problem.
```
#include <iostream>
#include <atomic>
#include <utility>
#include <cassert>
#include <thread>

using namespace std;

struct Counter {
  atomic<uint64_t> u64;
  atomic<uint64_t> avgcount;
  atomic<uint64_t> avgcount2;
  Counter() : u64(0), avgcount(0), avgcount2(0) { }
  std::pair<uint64_t, uint64_t> ReadValue() {
    uint64_t s, c;
    do {
      c = avgcount.load();
      s = u64.load();
    } while(avgcount2.load() != c);
   
    return make_pair(s, c);
  }

  void Inc(uint64_t v) {
    avgcount.fetch_add(1);
    u64.fetch_add(v);
    avgcount2.fetch_add(1);
  }
};

void test_func_inc(Counter *counter) {
  int i =10000000;
  while (i--) {
    // increase by one each time, so avgcount == u64 == avgcount2
    counter->Inc(1);
  }
  
}

void test_func_read(Counter *counter) {
  int i =10000000;
  while (i--) {
    std::pair<uint64_t, uint64_t> dat;
    dat = counter->ReadValue();

    if (dat.first != dat.second) {
      std::cout << "first " << dat.first << " count:" << dat.second << std::endl;
      assert(dat.first == dat.second);
    }    
  }
}


int main(int argc, char const *argv[]) {
  Counter counter;
  std::thread t1(test_func_inc, &counter);
  std::thread t2(test_func_read, &counter);
  t2.join();
  t1.join();
  return 0;
}
```
output:
[torch@torch-pc code]$ ./counter_test 
**first 250 count:249**
counter_test: counter_test.cc:48: void test_func_read(Counter*): Assertion `dat.first == dat.second' failed.
Aborted (core dumped)

**Comment by luhp** [comment]
@tchaikov 
I have retest the modified code, it's ok.

**Comment by tchaikov** [comment]
@luhp sorry for the confusion, i was instructing the jenkins to rerun `make check` to understand how often the build failure shows up on the arm64 builders.

your fix looks great! but before it gets merged, you might want to address following items:

- could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes , in this case, it's "common: ".
- 
could you add a "`Signed-off-by`" line at the end of your commit message? "git commit -s" will do the trick for you. see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#1-sign-your-work
- and a test in `src/test/perf_counters.cc` would be great.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-08-03_07:35:53-rados-wip-kefu-testing-2018-08-03-0109-distro-basic-smithi/

- some of the failures are caused by #23257
- the other are fixed by #23411 

@luhp ping?

File: src/test/perf_counters.cc [hunk 1]
```diff
@@ -219,3 +220,41 @@ TEST(PerfCounters, ResetPerfCounters) {
...
+    // increase by one, make sure data.u64 equal to data.avgcount
+    fake_pf->inc(TEST_PERFCOUNTERS3_ELEMENT_READ, 1);
+  }
+}
+
+static void counters_readavg_test(PerfCounters* fake_pf) {
+  int i = 10000;
+  while(i--) {
+    std::pair<uint64_t, uint64_t> dat = fake_pf->get_avg(TEST_PERFCOUNTERS3_ELEMENT_READ);
+    // if dat.first is not equal to dat.second, then we get a wrong data
```
Comment (tchaikov) [inline src/test/perf_counters.cc]: this comment is redundant. better off putting something like

> sum and count should be identical as we increment `TEST_PERFCOUNTERS3_ELEMENT_READ` by 1 everytime.

File: src/common/perf_counters.h [hunk 2]
```diff
@@ -233,6 +233,7 @@ class PerfCounters
   void dec(int idx, uint64_t v = 1);
   void set(int idx, uint64_t v);
   uint64_t get(int idx) const;
+  std::pair<uint64_t, uint64_t> get_avg(int idx) const;
```
Comment (tchaikov) [inline src/common/perf_counters.h]: strictly speaking, this change does not belong to the fix, i'd suggest move it into its own commit.

File: src/common/perf_counters.cc [hunk 3]
```diff
@@ -230,6 +230,22 @@ uint64_t PerfCounters::get(int idx) const
   return data.u64;
 }
 
+std::pair<uint64_t, uint64_t> PerfCounters::get_avg(int idx) const
+{
+  if (!m_cct->_conf->perf)
+    return make_pair(0, 0);
```
Comment (tchaikov) [inline src/common/perf_counters.cc]:
```c++
return {0, 0};
```
would suffice.

File: src/common/perf_counters.cc [hunk 4]
```diff
@@ -230,6 +230,22 @@ uint64_t PerfCounters::get(int idx) const
...
+
+  assert(idx > m_lower_bound);
+  assert(idx < m_upper_bound);
+  const perf_counter_data_any_d& data(m_data[idx - m_lower_bound - 1]);
```
Comment (tchaikov) [inline src/common/perf_counters.cc]: use `auto` if you please, like
```c++
auto& data = m_data[idx - m_lower_bound - 1];
```

**Comment by tchaikov** [comment]
@luhp 

- not all your commits have `Signed-off-by` line. could you add a "`Signed-off-by`" line at the end of your commit message? "git commit -s" will do the trick for you. see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#1-sign-your-work
- your `Fixes` is self-referencing this very PR, i'd suggest you use something like
```
Fixes: http://tracker.ceph.com/issues/25211 
```
- please fold related commits into a single one. and remove the "Merge" commit from your PR.

the title of your commit message is a little bit vague, IMHO.

> fixed bug in PerfCounters::perf_counter_data_any_d::read_avg

how about 

> fix a racing in PerfCounters::perf_counter_data_any_d::read_avg

?

**Comment by luhp** [comment]
@tchaikov  Thanks for your advise.

File: src/common/perf_counters.h (hunk 2 above)
Comment (tchaikov) [reply src/common/perf_counters.h]: i knew. but the reason you added it is actually for testing the fix for `get_tavg_ns()`. ideally, we should be testing `get_tavg_ns()` instead, no?

if there is no consumers of an API, we intend to avoid adding it.

Comment (luhp) [reply src/common/perf_counters.h]: Sorry, I did't notice that get_tavg_ns() exists already. I worked on jewel before.

**Comment by luhp** [comment]
@tchaikov  I don't know why make check (arm64) failed.

**Comment by tchaikov** [comment]
@luhp the test failure on aarch64 is a known issue.
